lint/rule_enabled.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 		https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lint
+
+import (
+	"strings"
+
+	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+)
+
+// Disable all rules for deprecated descriptors.
+func disableDeprecated(d desc.Descriptor) bool {
+	switch v := d.(type) {
+	case *desc.EnumDescriptor:
+		return v.GetEnumOptions().GetDeprecated()
+	case *desc.EnumValueDescriptor:
+		return v.GetEnumValueOptions().GetDeprecated()
+	case *desc.FieldDescriptor:
+		return v.GetFieldOptions().GetDeprecated()
+	case *desc.FileDescriptor:
+		return v.GetFileOptions().GetDeprecated()
+	case *desc.MessageDescriptor:
+		return v.GetMessageOptions().GetDeprecated()
+	case *desc.MethodDescriptor:
+		return v.GetMethodOptions().GetDeprecated()
+	case *desc.ServiceDescriptor:
+		return v.GetServiceOptions().GetDeprecated()
+	}
+	return false
+}
+
+// Provide methods that are able to disable rules.
+//
+// This pattern is a hook for internal extension, by creating an additional
+// file in this package that can add an additional check:
+//
+//    func disableForInternalReason(d desc.Descriptor) bool { ... }
+//
+//    func init() {
+//      descriptorDisableChecks = append(descriptorDisableChecks, disableForInternalReason)
+//    }
+var descriptorDisableChecks = []func(d desc.Descriptor) bool{
```

Comment:
Do you want to use a map with a key(name) to describe the disabling check function? So if we find a disabling check is not applicable internally, we can remove the check based on the key(name).

Reply:
I'd be up for making this change the first time it comes up, but for now I would prefer match [rule_urls.go](https://github.com/googleapis/api-linter/blob/master/lint/rule_urls.go) since that already exists and uses a slice.